### ISSUE TITLE
bug in origin_time_source_matcher

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#323]
Messed up somehow what I started this and got this empty box.  Actual comment is below.

**Comment by pavlis** [comment]
Messed up starting this somehow, but I found what is without doubt a bug in the normalization class `origin_time_source_matcher.   The following method will fail if the input is an ensemble:

```
    def _get_test_time(self, d, time):
        """
        Private method that safely computes a time for testing from 
        the "starttime" attribute of d.  
        """
        if time == None:
            if isinstance(
                d, (TimeSeries, Seismogram, TimeSeriesEnsemble, SeismogramEnsemble)
            ):
                test_time = d.t0 - self.t0offset
            else:
                if d.is_defined("starttime"):
                    test_time = d["starttime"] - self.t0offset
                else:  #   t0 can't be extracted from the object
                    return None
        else:
            test_time = time - self.t0offset
        return test_time
```
The problem is that `t0` is defined only for TimeSeries or Seismogram data.  An ensemble does not always have the concept of a start time.   I think the idea of this private method is to try to safely always retrieve a valid time for testing, but this will fail often if passed Time==None.   

We need this algorithm to work with ensembles as common source gathers are a very very common construct.   I think what we need is to add a time key as an optional argument to the constructor.  The logic should be something like this:
```
if self.time_key: 
   if d.is_defined(self.time_key):
     test_time = d[self.tim_key] - self.t0offset
   else:
     return None   # matches logic but may not be right approach here
else:
   if isinstance(d, (TimeSeries, Seismogram)):
      test_time = d.t0 - self.t0offset
   else:
     return None
return test_time
```
or something like that.

**Comment by wangyinz** [comment]
I got that fixed in #324 and will merge it soon after checks passed.
